### ISSUE TITLE
UnreachableCode false positives in 1.23.0

### ISSUE THREAD

**Issue by josephlbarnett** [detekt/detekt#6129]
## Expected Behavior

should not be any unreachable code reported in https://github.com/trib3/leakycauldron/blob/main/graphql/src/main/kotlin/com/trib3/graphql/resources/GraphQLSseResource.kt (or at least, none of this code looks obviously unreachable to me?)

## Observed Behavior

```
[INFO]    [detekt] /home/circleci/repo/graphql/src/main/kotlin/com/trib3/graphql/resources/GraphQLSseResource.kt:242:9: This expression is unreachable code which should either be used or removed. [UnreachableCode]
[INFO]    [detekt] /home/circleci/repo/graphql/src/main/kotlin/com/trib3/graphql/resources/GraphQLSseResource.kt:243:9: This expression is unreachable code which should either be used or removed. [UnreachableCode]
[INFO]    [detekt] /home/circleci/repo/graphql/src/main/kotlin/com/trib3/graphql/resources/GraphQLSseResource.kt:244:9: This expression is unreachable code which should either be used or removed. [UnreachableCode]
[INFO]    [detekt] /home/circleci/repo/graphql/src/main/kotlin/com/trib3/graphql/resources/GraphQLSseResource.kt:245:9: This expression is unreachable code which should either be used or removed. [UnreachableCode]
[INFO]    [detekt] /home/circleci/repo/graphql/src/main/kotlin/com/trib3/graphql/resources/GraphQLSseResource.kt:248:9: This expression is unreachable code which should either be used or removed. [UnreachableCode]
[INFO]    [detekt] /home/circleci/repo/graphql/src/main/kotlin/com/trib3/graphql/resources/GraphQLSseResource.kt:249:9: This expression is unreachable code which should either be used or removed. [UnreachableCode]
[INFO]    [detekt] /home/circleci/repo/graphql/src/main/kotlin/com/trib3/graphql/resources/GraphQLSseResource.kt:250:9: This expression is unreachable code which should either be used or removed. [UnreachableCode]
[INFO]    [detekt] /home/circleci/repo/graphql/src/main/kotlin/com/trib3/graphql/resources/GraphQLSseResource.kt:252:9: This expression is unreachable code which should either be used or removed. [UnreachableCode]
[INFO]    [detekt] /home/circleci/repo/graphql/src/main/kotlin/com/trib3/graphql/resources/GraphQLSseResource.kt:261:9: This expression is unreachable code which should either be used or removed. [UnreachableCode]
[INFO]    [detekt] /home/circleci/repo/graphql/src/main/kotlin/com/trib3/graphql/resources/GraphQLSseResource.kt:273:9: This expression is unreachable code which should either be used or removed. [UnreachableCode]
[INFO]    [detekt] /home/circleci/repo/graphql/src/main/kotlin/com/trib3/graphql/resources/GraphQLSseResource.kt:274:9: This expression is unreachable code which should either be used or removed. [UnreachableCode]
```

## Steps to Reproduce

run detekt 1.23 against this codebase
## Your Environment

* Version of detekt used: 1.23.0
* Version of Gradle used (if applicable): n/a, running in maven via antrun plugin
* Gradle scan link (add `--scan` option when running the gradle task):
* Operating System and version: ubuntu linux / circleci
* Link to your project (if it's a public repository): https://github.com/trib3/leakycauldron

**Comment by cortinico** [comment]
I think the offending line is:
```
val streamToken = headerToken ?: paramToken ?: throw WebApplicationException(unauthorizedResponse())
```

For some reason the "double" elvis operator is causing this rule to fail

**Comment by atulgpt** [comment]
Hi, @cortinico 

although below TC is passing
```
@Test
fun `does not reports unreachable code after nested elvis - #6129`() {
    val code = """
        fun test(
            opId: String,
            a: String?,
            b: String?
        ) {
            val finalString = a ?: b ?: throw RuntimeException("both null")
            println(opId + finalString)
        }
    """.trimIndent()
    assertThat(subject.compileAndLintWithContext(env, code)).isEmpty()
}
```

**Comment by cortinico** [comment]
Fair point. Then I have no idea why it's raising a failure after that line. We should investigate further

**Comment by marschwar** [comment]
I looked into this a little and found that the rule does not seem to work for types that are not known.

If we add this test for this snippet, it fails as well.

```kotlin
fun doSomething(a: UUID?, b: UUID?): Boolean {
    val streamToken = a ?: b ?: error()
    return true
}
```

Changing `UUID` to `String` will make the test pass.



@josephlbarnett I added `<arg value="--debug"/>` to your configuration an then I get the following message:

```
[INFO]    [detekt] error: supertypes of the following classes cannot be resolved. Please make sure you have the required dependencies in the classpath:
[INFO]    [detekt]     class graphql.GraphQLError, unresolved supertypes: java.lang.Object, java.io.Serializable
[INFO]    [detekt]     class org.eclipse.jetty.websocket.api.WebSocketAdapter, unresolved supertypes: java.lang.Object
[INFO]    [detekt]     class org.eclipse.jetty.websocket.api.WebSocketListener, unresolved supertypes: java.lang.Object
[INFO]    [detekt]     class org.eclipse.jetty.websocket.api.WebSocketConnectionListener, unresolved supertypes: java.lang.Object
[INFO]    [detekt]     class org.eclipse.jetty.websocket.core.server.WebSocketCreator, unresolved supertypes: java.lang.Object
...
[INFO]    [detekt]     class io.dropwizard.auth.AuthFilter, unresolved supertypes: java.lang.Object
[INFO]    [detekt]     class javax.ws.rs.container.ContainerRequestFilter, unresolved supertypes: java.lang.Object
[INFO]    [detekt]     class kotlin.Result, unresolved supertypes: java.io.Serializable
[INFO]    [detekt] Adding -Xextended-compiler-checks argument might provide additional information.

```

I have no idea why this worked with 1.22.0 but not in 1.23.0. Did we change how the classpath is passed to the CLI?

**Comment by marschwar** [comment]
It seems like the TC below starts to fail with [this commit](https://github.com/detekt/detekt/commit/43f0796c54a179bfd0e0e94c7c79128771bbbf2a). @3flex Do you have any idea how this could be related?

```kotlin
@Test
fun `does not reports unreachable code after nested elvis - #6129`() {
    val code = """
        import java.util.UUID

        fun doSomething(a: UUID?, b: UUID?): Boolean {
            val streamToken = a ?: b ?: error()
            return true
        }
    """.trimIndent()
    assertThat(subject.compileAndLintWithContext(env, code)).isEmpty()
}
```

**Comment by atulgpt** [comment]
Hi, @marschwar strange in my PC above TC is passing(I am in latest main)

**Comment by 3flex** [comment]
Most likely due to a Kotlin change which was worked around here: https://github.com/detekt/detekt/commit/406be3373614e6c876c435432f1ee602ff05e58a

Guess it doesn't work for Maven.

@josephlbarnett can you try running that command but with JAVA_HOME set?

I don't know if this is the root cause and I don't use Maven at all, but hopefully this is a push in the right direction.

**Comment by marschwar** [comment]
> Hi, @marschwar strange in my PC above TC is passing(I am in latest main)

That is very odd. Running `./gradlew :detekt-rules-errorprone:test --tests UnreachableCodeSpec` works fine and the tests pass. Running the tests in IntelliJ (using gradle to run the tests) has the test fail.

**Comment by atulgpt** [comment]
Hi, @marschwar have you set different JDK path for terminal and using integrated one in IntelliJ? I ran from IntelliJ green triangle version as well as terminal from both places it ran successfully. 
IntelliJ java version: openjdk 17.0.6 2023-01-17
Terminal JAVA_HOME = openjdk 17.0.6 2023-01-17

Against which java version you are running this TC?

**Comment by josephlbarnett** [comment]
> @josephlbarnett can you try running that command but with JAVA_HOME set?
> 
> I don't know if this is the root cause and I don't use Maven at all, but hopefully this is a push in the right direction.

still fails with JAVA_HOME set

**Comment by marschwar** [comment]
> Hi, @marschwar have you set different JDK path for terminal and using integrated one in IntelliJ? I ran from IntelliJ green triangle version as well as terminal from both places it ran successfully. IntelliJ java version: openjdk 17.0.6 2023-01-17 Terminal JAVA_HOME = openjdk 17.0.6 2023-01-17
> 
> Against which java version you are running this TC?

The same for terminal and IDE.

openjdk version "17.0.6" 2023-01-17
OpenJDK Runtime Environment Temurin-17.0.6+10 (build 17.0.6+10)
OpenJDK 64-Bit Server VM Temurin-17.0.6+10 (build 17.0.6+10, mixed mode, sharing)

That is very odd.

**Comment by yogurtearl** [comment]
Unreachable code is already a compiler warning in Kotlin. 
Do we need a detekt check for unreachable code as well?

**Comment by atulgpt** [comment]
Hi, @yogurtearl there can be cases when a compiler warning is not checked against CI while Detekt is. Also, there are cases where we have created the Detekt rule against compiler warning

**Comment by 3flex** [comment]
Is anyone willing & able to check whether the 2.0.0 alpha fixes this?

**Comment by josephlbarnett** [comment]
> Is anyone willing & able to check whether the 2.0.0 alpha fixes this?

fyi, tried and ran into #8626

**Comment by josephlbarnett** [comment]
commenting to keep alive.

**Comment by josephlbarnett** [comment]
still a problem

**Comment by atulgpt** [comment]
HI @josephlbarnett Is this still a problem with latest alpha?

**Comment by josephlbarnett** [comment]
unable to run the latest alpha per https://github.com/detekt/detekt/issues/8626

**Comment by josephlbarnett** [comment]
bump

**Comment by atulgpt** [comment]
Hi @josephlbarnett can you test the code using the hack mentioned at comment https://github.com/detekt/detekt/issues/8626#issuecomment-3608824612 and then try with the latest alpha version?

**Comment by josephlbarnett** [comment]
@atulgpt using that hack, and enabling UnreachableCode, I do not get any errors.
BUT enabling other rules that I know my codebase violates (eg, SpreadOperator, LongParameterList with lower thresholds, TooManyFunctions w/ ignorePrivate disabled) also does not generate any errors.  Some rules do seem to work as intended though (ForbiddenComment, TooGeneriExceptionCaught), so not sure if this means the bug is fixed or if UnreachableCode is not executing/working properly (I'm pretty sure the code that generated this issue hasn't changed at least)?

**Comment by atulgpt** [comment]
> @atulgpt using that hack, and enabling UnreachableCode, I do not get any errors.
> BUT enabling other rules that I know my codebase violates (eg, SpreadOperator, LongParameterList with lower thresholds, TooManyFunctions w/ ignorePrivate disabled) also does not generate any errors.  Some rules do seem to work as intended though (ForbiddenComment, TooGeneriExceptionCaught), so not sure if this means the bug is fixed or if UnreachableCode is not executing/working properly (I'm pretty sure the code that generated this issue hasn't changed at least)?

Hi thanks for the clarification, for testing if unreachable rule works or not, you can add a violation and then test if that violation is getting reported or not 

Also remember that previously(in 1.x.x versions), some rule ran in both mode, with type resolution and without type resolution detekt tasks. In other words, if you just run `detekt` task(`detekt` plain task runs without type resolution), then earlier some rule reported few er violations then same rules running in type resolution. While  now in 2.x.x versions, those rules will be completely disabled. Rule requiring type resolution will only run in variant specific detekt tasks like `detektMain` or `detektDebug`

For other issues can you create seperate issues

If you conclude that unreachable rule is working fine then you can close this issue

**Comment by josephlbarnett** [comment]
ok yeah if I add an early return statement the UnreachableCode rule still "passes" so it must not be executing/working properly...

**Comment by josephlbarnett** [comment]
are "variant specific detekt tasks like detektMain or detektDebug" gradle specific or available @ cmdline? docs seem to say that as long as --classpath and --jvm-target are provided it should work?
